ipld/amt/src/iter.rs

```diff
@@ -0,0 +1,559 @@
+                        }
+                    }
+                    iter.stack.pop();
+                }
+                Some(Node::Link { links }) => {
+                    if stack.idx < links.len() {
+                        let link = &links[stack.idx];
+                        match link {
+                            Some(Link::Cid { cid, cache }) => {
+                                match cache.get_or_try_init(|| {
+                                    iter.blockstore
+                                        .get_cbor::<CollapsedNode<V>>(cid)?
+                                        .ok_or_else(|| Error::CidNotFound(cid.to_string()))?
+                                        .expand(iter.bit_width)
+                                        .map(Box::new)
+                                }) {
+                                    Ok(node) => {
+                                        stack.idx += 1;
+                                        iter.stack.push(IterStack {
+                                            node: Some(node.as_ref()),
+                                            idx: 0,
+                                        });
+                                    }
+                                    Err(e) => return Err(e),
+                                }
+                            }
+                            Some(Link::Dirty(node)) => {
+                                stack.idx += 1;
+                                iter.stack.push(IterStack {
+                                    node: Some(node.as_ref()),
+                                    idx: 0,
+                                });
+                            }
+                            None => {
+                                stack.idx += 1;
+                                iter.key += 2_usize.pow(iter.bit_width);
+                            }
+                        };
+                    } else {
+                        iter.stack.pop();
+                    }
+                }
+                None => {
+                    return Ok(Iter {
+                        stack: vec![],
+                        blockstore: &self.block_store,
+                        bit_width: self.bit_width(),
+                        ver: PhantomData,
+                        key: 0,
+                    })
+                }
+            }
+        }
+        Ok(iter)
+    }
+}
+
+impl<'a, V, BS, Ver> IntoIterator for &'a crate::AmtImpl<V, BS, Ver>
+where
+    V: Serialize + DeserializeOwned,
+    Ver: crate::root::version::Version,
+    BS: Blockstore,
+{
+    type IntoIter = Iter<'a, V, &'a BS, Ver>;
+    type Item = Result<(usize, &'a V), crate::Error>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+pub struct Iter<'a, V, BS, Ver> {
+    stack: Vec<IterStack<'a, V>>,
+    blockstore: BS,
+    bit_width: u32,
+    ver: PhantomData<Ver>,
+    key: usize,
+}
+
+pub struct IterStack<'a, V> {
+    pub(crate) node: Option<&'a Node<V>>,
```

Comment:
Does this need to be an `Option`?

Reply:
At some (pun intended) point I think it did, but with changes since then, it doesn't need to be any more. Fixed.